README.md
# mush

Demonstrates benchmarking in Clojure

## Usage

`lein test`

Look in the benchmarks directory for output files

## License

Copyright © 2016 Timothy Pratley

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-01-30
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-01-30
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/mush/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/mush/compare/0.1.0...0.1.1


test/mush/core_test.clj
(ns mush.core-test
  (:require
    [clojure.test :refer :all]
    [clojure.java.io :as io]
    [clojure.string :as string]
    [mush.benchmarking :as b]
    [mush.core :as mush]
    [criterium.core :as cr]
    [clj-jgit.porcelain :as porc]
    [clj-jgit.querying :as q]
    [camel-snake-kebab.core :as kebab]))

(use-fixtures :once b/save-gathered-benchmarks)

(b/bench-press "stuff" (* 1000 b/nsecs)
               (mush/foo 1))


test/mush/benchmarking.clj
(ns mush.benchmarking
  (:require
    [clojure.test :refer :all]
    [clojure.java.io :as io]
    [clojure.string :as string]
    [criterium.core :as cr]
    [clj-jgit.porcelain :as porc]
    [clj-jgit.querying :as q]
    [camel-snake-kebab.core :as kebab]))

(defn format-time [t]
  (apply cr/format-value t (cr/scale-time t)))

(def nsecs 1e9)
(def msecs 1000)

(def git-commit-info
  (porc/with-repo "." (some-> (q/rev-list repo)
                              (first)
                              (->> (q/commit-info repo))
                              (dissoc :raw :repo))))

(def benchmarks
  (atom {(:time git-commit-info) {:git-commit-info git-commit-info}}))

(defn add-results! [k results]
  (swap! benchmarks assoc-in [(:time git-commit-info) k] results))

(def benchmark-filename
  (str (string/replace (:time git-commit-info) " " "_")
       "-" (:id git-commit-info) ".edn"))

(defn save-gathered-benchmarks [f]
  (f)
  (with-open [w (io/writer (doto (io/file "benchmarks" benchmark-filename)
                             (io/make-parents)))]
    (clojure.pprint/pprint @benchmarks w)))

(defmacro bench-press [title acceptable-time & form]
  `(deftest ~(symbol (kebab/->kebab-case title))
     (testing ~title
       (let [results# (assoc (cr/quick-benchmark ~@form nil)
                        :acceptable-time ~acceptable-time)]
         (add-results! ~title results#)
         (is (< (first (:mean results#)) ~acceptable-time))))))

(defmacro time-msecs
  "Returns the time in msecs taken to evaluate expr."
  [expr]
  `(let [start# (. System (nanoTime))]
     ~expr
     (/ (double (- (. System (nanoTime)) start#)) 1000000.0)))


test/mush/simulation_test.clj
(ns mush.simulation-test
  (:require
    [clojure.test :refer :all]
    [mush.simulation :refer-all]
    [mush.benchmarking :as b]))

(use-fixtures :once b/save-gathered-benchmarks)

(deftest slow-tests
  (is (< (b/time-msecs (sim-project project1))
         1000)
      "A full project simulation should take less than one second"))

(deftest bad-microbenchmark
  (time (inc 1)))

(b/bench-press "Remove 1 done story" (* 1000 b/nsecs)
               (remove-done-stories project1))


src/mush/core.clj
(ns mush.core
  (:require [clojure.string :as string])
  (:import (java.util Date)))

(def project1
  {:title "Build a task tracking solution"
   :team '("timothypratley@gmail" "jeremy" "x@y")
   :stories '({:id "aaa"
               :title "Build a ClojureScript front end"
               :status "Ready"
               :as-of (Date.)}
               {:id "bbb"
                :title "Build backend services"
                :status "Done"
                :as-of (Date.)}
               {:id "ccc"
                :title "Tune performance"
                :status "In Progress"
                :as-of (Date.)})})

(def buzz-words
  ["create" "implement" "refactor" "tune" "resize" "load" "verify" "update"
   "widget" "feature" "wizard" "sorting" "search" "crud" "ui" "dashboard" "history" "configuration" "dropdown"
   "deploy" "password" "authentication" "performance" "monitoring" "stream" "social" "flat" "experience"
   "and" "or" "but" "with" "then" "that" "who" "why" "when" "because" "a"
   "firefox" "ie" "osx" "windows" "disk" "space" "crash" "freeze" "blank"])

(defn fetch-history [member]
  (Thread/sleep 10)
  {:some "data"})

(defn fetch-profile [member]
  (Thread/sleep 11)
  {:more "data"})

(defn member-info [member]
  (let [history (fetch-history member)
        profile (fetch-profile member)]
    (merge history profile)))

(defn stories-by-status [{:keys [stories]} status]
  (filter #(= (:status %) status) stories))

(defn unique []
  (Integer/toString (rand-int (Math/pow 36 3)) 36))

(defn add-story [project]
  (update project :stories conj
          {:id (unique)
           :title (string/capitalize (string/join " " (take (+ 3 (rand-int 9)) (shuffle buzz-words))))
           :status "Ready"
           :as-of (Date.)}))

(defn remove-story [project story]
  (update project :stories
          (fn [stories]
            (remove #(= (:id %) (:id story)) stories))))

(defn update-story-status [project story status]
  (update project :stories
          (fn [stories]
            (for [s stories]
              (if (= (:id s) (:id story))
                (assoc s :status status)
                s)))))

(defn discover-new-stories [project]
  (nth (iterate add-story project) (rand-int 2)))

(defn start-story [project story]
  (-> project
    (update-story-status (:id story) "In Progress")
    (discover-new-stories)))

(defn start-stories [project]
  (let [team-count (count (:team project))
        doing-count (count (stories-by-status project "In Progress"))]
    (reduce start-story project
            (take (- team-count doing-count) (stories-by-status project "Ready")))))

(defn completion-rate [project]
  (reduce (fn [acc member]
            (+ acc (:stories-per-day (member-info member) 0.5)))
          0
          (:team project)))

(defn complete-stories [project]
  (reduce (fn complete-story [acc story]
            (update-story-status acc story "Done"))
          project
          (take (rand-int (int (* (completion-rate project) 2)))
                (stories-by-status project "In Progress"))))

(defn standup [project]
  (-> project
    (complete-stories)
    (start-stories)))

(defn remove-done-stories [project]
  (reduce remove-story project (stories-by-status project "Done")))

(defn new-requirements [project]
  (nth (iterate add-story project) (count (:team project))))

(defn prioritize [project]
  (reduce remove-story project
          (drop (* 3 (count (:team project)))
                (shuffle (stories-by-status project "Ready")))))

(defn planning [project]
  (-> project
    (remove-done-stories)
    (new-requirements)
    (prioritize)))

(defn sim-week [project]
  (planning (nth (iterate standup project) 5)))

(defn sim-project [project]
  (nth (iterate sim-week project) 20))


(defn dont-use-lists []
  )

(defn lots-of-reflection []
  )


src/mush/simulation.clj
(ns mush.simulation
  "Many of these functions are intentionally slow to act as examples for benchmarking and performance tuning options."
  (:require [clojure.string :as string])
  (:import (java.util Date)))

(def project1
  {:title "Build a task tracking solution"
   :team '("timothypratley@gmail" "jeremy" "x@y")
   :stories '({:id "aaa"
               :title "Build a ClojureScript front end"
               :status "Ready"
               :as-of (Date.)}
               {:id "bbb"
                :title "Build backend services"
                :status "Done"
                :as-of (Date.)}
               {:id "ccc"
                :title "Tune performance"
                :status "In Progress"
                :as-of (Date.)})})

(defn stories-by-status [{:keys [stories]} status]
  (filter #(= (:status %) status) stories))

(def buzz-words
  ["create" "implement" "refactor" "tune" "resize" "load" "verify" "update"
   "widget" "feature" "wizard" "sorting" "search" "crud" "ui" "dashboard" "history" "configuration" "dropdown"
   "deploy" "password" "authentication" "performance" "monitoring" "stream" "social" "flat" "experience"
   "and" "or" "but" "with" "then" "that" "who" "why" "when" "because" "a"
   "firefox" "ie" "osx" "windows" "disk" "space" "crash" "freeze" "blank"])

(defn unique []
  (Integer/toString (rand-int (Math/pow 36 3)) 36))

(defn add-story [project]
  (update project :stories conj
          {:id (unique)
           :title (string/capitalize (string/join " " (take (+ 3 (rand-int 9)) (shuffle buzz-words))))
           :status "Ready"
           :as-of (Date.)}))

(defn remove-story [project story]
  (update project :stories
          (fn [stories]
            (remove #(= (:id %) (:id story)) stories))))

(defn update-story-status [project story status]
  (update project :stories
          (fn [stories]
            (for [s stories]
              (if (= (:id s) (:id story))
                (assoc s :status status)
                s)))))

(defn discover-new-stories [project]
  (nth (iterate add-story project) (rand-int 2)))

(defn start-story [project story]
  (-> project
    (update-story-status (:id story) "In Progress")
    (discover-new-stories)))

(defn start-stories [project]
  (let [team-count (count (:team project))
        doing-count (count (stories-by-status project "In Progress"))]
    (reduce start-story project
            (take (- team-count doing-count) (stories-by-status project "Ready")))))

(defn fetch-history [member]
  (Thread/sleep 10)
  {:some "data"})

(defn fetch-profile [member]
  (Thread/sleep 11)
  {:more "data"})

(defn member-info [member]
  (let [history (fetch-history member)
        profile (fetch-profile member)]
    (merge history profile)))

(defn member-info-better [member]
  (let [history (future (fetch-history member))
        profile (future (fetch-profile member))]
    (merge @history @profile)))

(defn completion-rate [project]
  (reduce (fn [acc member]
            (+ acc (:stories-per-day (member-info member) 0.5)))
          0
          (:team project)))

(def completion-rate-better
  (memoize completion-rate))

(defn complete-stories [project]
  (reduce (fn complete-story [acc story]
            (update-story-status acc story "Done"))
          project
          (take (rand-int (int (* (completion-rate project) 2)))
                (stories-by-status project "In Progress"))))

(defn complete-stories-alternative [project]
  (let [rate (completion-rate project)]
    (reduce (fn complete-story [acc story]
              (update-story-status acc story "Done"))
            project
            (take (rand-int (int (* rate 2)))
                  (stories-by-status project "In Progress")))))

(defn standup [project]
  (-> project
    (complete-stories)
    (start-stories)))

(defn remove-done-stories [project]
  (reduce remove-story project (stories-by-status project "Done")))

(defn new-requirements [project]
  (nth (iterate add-story project) (count (:team project))))

(defn prioritize [project]
  (reduce remove-story project
          (drop (* 3 (count (:team project)))
                (shuffle (stories-by-status project "Ready")))))

(defn planning [project]
  (-> project
    (remove-done-stories)
    (new-requirements)
    (prioritize)))

(defn sim-week [project]
  (planning (nth (iterate standup project) 5)))

(defn sim-week-alternative [project]
  (with-redefs [member-info (memoize member-info)]
    (sim-week project)))

(defn sim-project [project]
  (nth (iterate sim-week project) 20))




(defn dont-use-lists []
  )

(defn lots-of-reflection []
  )

(defn remove-done-stories-with-transient [project]
  (let [p (transient projects)]
    (persistent!
      (reduce (fn [acc story]
                (assoc! acc :stories
                          (fn [stories]
                            (remove #(= (:id %) (:id story)) stories))))
              p
              (stories-by-status project "Done")))))